### DISCUSSION TITLE
[Proposal]: [Dedicated case for invalid value in the switch statement and expression]

### DISCUSSION THREAD

**Discussion by tymtam2** [dotnet/csharplang#4577] [Language Ideas]
# Dedicated case for invalid value in the switch statement and expression

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Introduce special cases for the swtich statement and switch expression (`unmapped` and `?` respectively) for enums
to
handle unmapped values (as opposed to all "other mapped values") 
in order to
keep compile time ability to detect unhandled known values
because 
the current `default`/`_` solution removes compile time detection of known unhandled valus in order to handle the fact that c# enums can have invalid values. 

For example:
```
public enum Directions { Up, Down, Right, Left }
    
(...)    
var o = direction switch
{
     Directions.Up    => "vertical",
     Directions.Down => "vertical",
     _ => "horizontal",
     ? => throw new Exception($"Invalid value for direction: {direction}");
};
```

## Motivation
[motivation]: #motivation

Enums in C# are special. A variable of an enum type can hold a value that are no mapped in the enum declaration. The switch expression's and switch statement's abiltiy to detect errors at compile time is compromised because developers end up adding `_` or `default` to handle umapped values and loose the compile time warning about unhandled mapped values. Providing a dedicated case for unhandled values will create a safer, more correct code. 

## Detailed design
[design]: #detailed-design

Introduce `unmapped` for the switch statement and `?` for the switch expression applicable when switching on an enum. The `unmapped`/`_` case is invoked when the examined value is not mapped. 

Let's consider the following setup.
```
public enum Directions { Up, Down, Right, Left }
    
(...)    
var o = direction switch
{
     Directions.Up    => "vertical",
     Directions.Down => "vertical",
     Directions.Right    => "horizonal",
     Directions.Left => "horizontal",
     ? => throw new Exception($"Invalid value for direction: {direction}");
};
```

If `direction` is (Directions)5 then `?` is invoked. If the value is `Directions.Left` then `_` is invoked. 

If `?` is not provided the `_` is invoked. 

The primary use case is keeping the ability to detect unhandled cases when adding new values to enum.
```
public enum Directions { Up, Down, Right, Left, Forward, Back  } 
    
(...)    
var o = direction switch
{
     Directions.Up    => "vertical",
     Directions.Down => "vertical",
     Directions.Right    => "horizonal",
     Directions.Left => "horizontal",
     ? => throw new Exception($"Invalid value for direction: {direction}");
     // Compile time warning: Forward, Back are not handled 
};
```

Currently the default handler the error of missing handlers would not be detected until runtime:
```
public enum Directions { Up, Down, Right, Left, Forward, Back  }
    
(...)    
var o = direction switch
{
     Directions.Up    => "vertical",
     Directions.Down => "vertical",
     Directions.Right    => "horizonal",
     Directions.Left => "horizontal",
     _ => throw new Exception($"Invalid value for direction: {direction}"); // Only at runtime. So late!
};
```

## Drawbacks
[drawbacks]: #drawbacks

Code is not backward compatible. The code may slow down. 

## Alternatives
[alternatives]: #alternatives

Allow first calls, in code, acknowledment of 
`Warning CS8509 The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '...' is not covered.`

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by YairHalberstadt** [comment]
I think this is definitely an interesting idea. There are some things which will need clarification:

1. Currently everything acceptable in a switch expression is also a valid pattern. So I think this should be a pattern rather than specifically for a switch expression.
2. What should the codegen be for this?
3. How should this handle flags enums?
4. Are we looking at unmapped values at compile time or runtime? I presume compile time, but that should ideally be called out explicitly in the proposal.

Finally DUs present a better solution to this whole problem. Is this worth doing, or should we just wait till they get implemented.

Also if this could be written in 'specese' that would be useful. Id be happy to help with that if you're not sure how to go about that.

**Reply by tymtam2** [reply]
@YairHalberstadt It took a while, but can work on this now. How and where do I start on the 'specese' version? 

Are the RE1 comments below a good start?

**Reply by tymtam2** [reply]
RE 4. Are we looking at unmapped values at compile time or runtime? I presume compile time, but that should ideally be called out explicitly in the proposal.

Yes, compile time! I mentioned it in "in order to keep compile time ability to detect unhandled known values".

**Comment by CyrusNajmabadi** [comment]
Converted this to a discussion :)

**Comment by tymtam2** [comment]
RE 1.   Currently everything acceptable in a switch expression is also a valid pattern. So I think this should be a pattern rather than specifically for a switch expression.


### Option 1 (Positive)

#### Declared Value Pattern

```antlr
declared_value_pattern
    : Type'!'
    ;
```

An expression *e* matches the pattern `Type'!'` when `Type.IsEnum` is not `true` or when `Type.IsEnum` is `true` and *e* is not in the set of declared values. 

Example 1: 

```csharp
public enum Directions { 
    Up = 1, 
    Down = 2, 
} 

var a = Directions.Up is Directions!; // true 
var c = 1 is Directions!; // true 
var d = 3 is Directions!; // false
```

Example 2

```csharp
[Flags]
public enum DirectionFlags { 
    Up = 1, 
    Down = 2, 
} 

var a2 = DirectionFlags.Up is DirectionsFlags!; // true
var b2 = (DirectionFlags.Up | DirectionFlags.Down) is DirectionsFlags!; // true 
var c2 = 1 is DirectionFlags!; // true 
var d2 = 3 is DirectionFlags!; // true
var e2 = 4 is DirectionFlags!; // false
```

---- 

Usage: 

```csharp
public enum Directions { 
    Up = 1, 
    Down = 2, 
} 

var o1 = direction switch
{
     Directions.Up => "vertical",
     not Directions! => throw new Exception($"Invalid value for direction: {direction}");
     // or 
     // var x when x is not Directions!  => throw new Exception($"Invalid value for direction: {x}");
     
    //  warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern 'Directions.Down' is not covered.
};

var o2 = direction switch
{
     Directions.Up => "vertical",
     Directions.Down => "vertical",
     not Directions! => throw new Exception($"Invalid value for direction: {direction}");

     // All values handled, no warning, but if Directions ever grows we're protected. 
};
```

**Reply by tymtam2** [reply]
RE 1.   Currently everything acceptable in a switch expression is also a valid pattern. So I think this should be a pattern rather than specifically for a switch expression.

### Option 2 (Negative)

#### Undeclared Value Pattern

```antlr
undeclared_value_pattern
    : '!'Type
    ;
```

An expression *e* matches when `Type.IsEnum` is `true` and *e* is not in the set of declared values.

For example:

```csharp
[Flags]
public enum Directions { 
    Up = 1, 
    Down = 2, 
} 

var a = Directions.Up is !Directions; // false 
var c = 1 is !Directions; // false 
var d = 3 is !Directions; // true
```

and 

```csharp
[Flags]
public enum DirectionFlags { 
    Up = 1, 
    Down = 2, 
} 

var a2 = DirectionFlags.Up is !DirectionsFlags; // false
var b2 = (DirectionFlags.Up | DirectionFlags.Down) is !DirectionsFlags; // false
var c2 = 1 is !DirectionFlags; // false
var d2 = 3 is !DirectionFlags; // false
var e2 = 4 is !DirectionFlags; // true
```

----

Usage: 

```csharp
public enum Directions { 
    Up = 1, 
    Down = 2, 
} 

var o1 = direction switch
{
     Directions.Up => "vertical",
     !Directions => throw new Exception($"Invalid value for direction: {direction}");
     // or 
     // var x when x is !Directions  => throw new Exception($"Invalid value for direction: {x}");
     
    //  warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern 'Directions.Down' is not covered.
};

var o2 = direction switch
{
     Directions.Up => "vertical",
     Directions.Down => "vertical",
     !Directions => throw new Exception($"Invalid value for direction: {direction}");

     // All values handled, no warning, but if Directions ever grows we're protected. 
};
```

**Comment by tymtam2** [comment]
RE 2. What should the codegen be for this?

I'm out of my depth here. I don't know what an acceptable codegen is.

Could you help me find out how *warning CS8509* in the following code is is generated? 

```csharp

SpecialNumbers x = (SpecialNumbers)271;

var s = x switch 
//  warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). 
// For example, the pattern 'SpecialNumbers.g' is not covered.
{
    SpecialNumbers.π => "Is there a pi in a pipeline?",
};

public enum SpecialNumbers { 
    π = 314, 
    g = 981, 
}
```

**Comment by tymtam2** [comment]
RE 3. How should this handle flags enums?

I think this could work for flags as well.

```csharp
var o = direction switch
{
     Directions.Up => "vertical",
     Directions.Down => "vertical",
     !Directions  => throw new Exception($"Invalid value: {color}") // this catches everything except 1,2, and 3
     // warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern 'Directions.Up | Directions.Down' is not covered.
or 
    // warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '3' is not covered.

};


[Flags]
public enum Directions { 
    Up = 1, 
    Down = 2, 
} 
```

**Comment by tymtam2** [comment]
RE: 5 Finally DUs present a better solution to this whole problem. Is this worth doing, or should we just wait till they get implemented.

Discriminated Unions are of course amazing and - for example - compile time support for them in Rust is great. I think they will prevent many bugs.

However, it will take some time to the world move away from classic enums and it's unlikely classic enums will be removed fully. 

Bugs from forgetting to extend a switch after adding a new enum value are just so sad. The proposed pattern would eliminate them (if components are recompiled :)). I believe this feature would make a difference. I strongly believe that `_ =>` is used to **only** get rid of the warning and this removes a significant portion of what we love about c# - compile time safety.

**Reply by TahirAhmadov** [reply]
I think DUs are a whole different feature, and .NET enums are here to stay forever, because they are useful for existing scenarios.

The solution, I think, is runtime enforcement of enum validity: basically, a check is added whenever a backing value, like an `int`, is "cast" to `enum` type that is marked either with a keyword or an attribute.

**Reply by HaloFour** [reply]
> The solution, I think, is runtime enforcement of enum validity: basically, a check is added whenever a backing value, like an int, is "cast" to enum type.

That would be a massively breaking change.

**Reply by TahirAhmadov** [reply]
@HaloFour oh yes, to clarify - it would be a new opt-in feature rather than modifying all existing `enum` types. Think an attribute or a keyword, like `closed enum Foo { ... }` or `[Closed] enum Foo { ... }`

**Reply by tymtam2** [reply]
> The solution, I think, is runtime enforcement of enum validity: basically, a check is added whenever a backing value, like an int, is "cast" to enum type that is marked either with a keyword or an attribute.

This looks like a different solution to what I'm proposing. 

The spirit of my proposals is to address two interrelated issues 
* preventing developers putting '_' in any enum-bases switch to eliminate the warning about unhandled values
* no compile time support catching not updated switches that don't handle new enum values

**Reply by TahirAhmadov** [reply]
@tymtam2 in a future world where it's impossible to have invalid enum values (for properly decorated `enum` types), your proposal would no longer be needed; and `_` will in effect become the same as `?` in your proposal.

**Reply by tymtam2** [reply]
> your proposal would no longer be needed; and _ will in effect become the same as ? in your proposal.

You're saying that for a new type that doesn't yet exist proposal is not needed. That's not fair. The proposal is for an existing core language type that's used in virtually any c# project.

**Reply by TahirAhmadov** [reply]
Yes, if your proposal is implemented, existing `enum`s get this cool new feature. But, if we get the ability to define `closed enum` (or `strict enum`), which prevents setting the value to any invalid backing value, I envision the overwhelming majority of `enum`s to be changed pretty quickly. At that point, the old `enum` types will be almost never needed and remain largely for legacy reasons. The benefit of that improvement (over just the `switch` like you suggested) is that it's a sweeping improvement which prevents bugs across the board, think `if` statements, dictionaries, etc.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> But, if we get the ability to define `closed enum` (or `strict enum`), which prevents setting the value to any invalid backing value, I envision the overwhelming majority of `enum`s to be changed pretty quickly.

Maybe in devs' own projects, but I doubt that any major framework, including the BCL, would consider such a change given the potential it has for breaking existing projects.

**Reply by TahirAhmadov** [reply]
> I doubt that any major framework, including the BCL, would consider such a change given the potential it has for breaking existing projects.

I can see how bad code can be broken if an existing BCL `enum` is changed to be `closed`. That does present a big challenge - how to measure the potential negative impact, what the strategy should be, etc.
However, these framework `enum`s also don't change frequently; I doubt there is a frequent need to handle the specific case of "new options added to a BCL enum".

**Reply by HaloFour** [reply]
@TahirAhmadov 

> I can see how bad code can be broken if an existing BCL `enum` is changed to be `closed`. That does present a big challenge - how to measure the potential negative impact, what the strategy should be, etc.

The strategy will be to leave them alone.  It's not a question of whether those enums have ever changed, it's a question of how much existing code may depend on their existing behavior, regardless of whether or not it would be "bad code".

**Reply by TahirAhmadov** [reply]
That strategy would be unacceptable to me from the perspective of losing the major benefit of introducing such a limitation. BCL has breaking changes with almost every major version and I don't see why we can't gradually change these enums to "closed" over time, granted it's communicated clearly.
I brought up the changing of such framework enums in the context of this proposal; here, the OP asks for a way to handle the case of an enum expression returning a value which is a valid option, but one which isn't expected yet. My point was, even if these framework enums are not changed to be "closed" (like you wrote), it's still extremely unlikely to have a need for the proposal in the OP to handle these framework enums.

**Reply by HaloFour** [reply]
> That strategy would be unacceptable to me from the perspective of losing the major benefit of introducing such a limitation. BCL has breaking changes with almost every major version and I don't see why we can't gradually change these enums to "closed" over time, granted it's communicated clearly.

I'd suggest that the team(s) would be much more concerned about the potential break to the billions of lines of existing source code written over the past 20 years.  The runtime team routinely rejects proposals that may impact existing code even substantially less than this one, and there is no direct line of communication to the infinite number of internal business projects that would be directly impacted, most likely unexpectedly.

**Reply by TahirAhmadov** [reply]
PS. To clarify, I'm not saying all framework enums, be it BCL or other libraries, absolutely have to change; it may be that the added cost of "closed" check is too much of a hit performance-wise and it's deemed that certain enums are better left as is. Also, it may be that "closed" is not a runtime check, but rather a compilation check only: for example, converting a backing value is not allowed at the language level. It needs to be thought through.

**Reply by tymtam2** [reply]
Closed enums sound like as very separate feature. I'm skeptical about people changing the type of an enum in projects. 

This proposal is more than "a new cool feature". At the moment developers put `_ => ` to get rid of a warning because there is no other choice, even though they don't really want `_`. 

I think that without this feature switch for enums is not that great: `_ => ` moves errors from compile time to runtime, and they can be caught at compile time.
